runtime/nvqir/stim/StimCircuitSimulator.cpp

```diff
@@ -0,0 +1,192 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "nvqir/CircuitSimulator.h"
+#include "nvqir/Gates.h"
+#include "stim.h"
+
+#include <bit>
+#include <iostream>
+#include <set>
+#include <span>
+
+using namespace cudaq;
+
+namespace nvqir {
+
+/// @brief The StimCircuitSimulator implements the CircuitSimulator
+/// base class to provide a simulator delegating to the Stim library from
+/// https://github.com/quantumlib/Stim.
+class StimCircuitSimulator : public nvqir::CircuitSimulatorBase<double> {
+protected:
+  stim::Circuit stimCircuit;
+  std::mt19937_64 randomEngine;
+
+  /// @brief Grow the state vector by one qubit.
+  void addQubitToState() override { addQubitsToState(1); }
+
+  /// @brief Override the default sized allocation of qubits
+  /// here to be a bit more efficient than the default implementation
+  void addQubitsToState(std::size_t qubitCount,
+                        const void *stateDataIn = nullptr) override {
+    return;
```

Comment:
nit: should we throw an error here when `stateDataIn` is not null?

Reply:
Updated in 20343fd07